### DISCUSSION TITLE
Has anyone tried to encrypt a certificate for MS Graph subscription?

### DISCUSSION THREAD

**Discussion by mohammadamer** [microsoft/hack-together#110] [Q&A]
Unfortunately, it's not clear in [MS documentation](https://learn.microsoft.com/en-us/graph/teams-changenotifications-chatmessage) how to encrypt a certificate for MS Graph Subscription.
Has anyone tried to encrypt a certificate for MS Graph subscription?

`             // Get ngrok URL if set (for local development)
                var notificationHost = _config["ngrokUrl"] ?? req.Url.Host;
                var certificateThumbprint = _config["CertificateThumbprint"];

                
                // Load the X509Certificate and add it to the subscription object
                var certificate = Utilities.X509CertificateUtility.LoadCertificate(StoreName.My, StoreLocation.CurrentUser, certificateThumbprint);
                var encryptionCertificateId = !string.IsNullOrEmpty(certificate?.FriendlyName) ? certificate.FriendlyName : certificate?.Subject;
                var encryptionCertificate = Convert.ToBase64String(certificate.Export(X509ContentType.Cert));

                // Create a new subscription object
                var subscription = new Subscription
                {
                    ChangeType = "created",
                    IncludeResourceData = true,
                    NotificationUrl = $"{notificationHost}/api/Notify",
                    Resource = "/teams/getAllMessages",// all messages and replies across channels.
                    //Resource = "/chats/getAllMessages",//all messages across chats in a tenant.
                    ExpirationDateTime = DateTime.UtcNow.AddMinutes(60),
                    ClientState = SetSubscription.ClientState,
                    EncryptionCertificateId = encryptionCertificateId,
                    EncryptionCertificate = encryptionCertificate,
                };
                //subscription.AddPublicEncryptionCertificate(certificate);

                _logger.LogInformation($"Creating subscription...");

                // POST /subscriptions
                var createdSubscription = await graphClient.Subscriptions.PostAsync(subscription);`

**Comment by waldekmastykarz** [comment]
Certificate itself doesn't have to be encrypted, just base64 encoded. Do you have some code that we could look like to better understand what you're trying and where you're stuck?

**Reply by mohammadamer** [reply]
Yep, sorry for miss explaining. just base64 encoded the certificate. This is what I'm trying to figure out.

fortunately, I found out that @andrueastman is using something like that it didn't work. This extension method is not present.

`subscription.AddPublicEncryptionCertificate(this._certificate);`

**Reply by waldekmastykarz** [reply]
Which version of the Microsoft Graph .NET SDK are you using: v4 or v5?

**Reply by mohammadamer** [reply]
Microsoft Graph .NET v5

**Reply by waldekmastykarz** [reply]
Let me check with @andrueastman

**Reply by mohammadamer** [reply]
When I change to Microsoft Graph .NET v4, Then I managed to encrypt the certificate. 
Now, I'm getting issue with decrypt the notification content.

`Message: Internal.Cryptography.CryptoThrowHelper.WindowsCryptographicException: 'Access denied.'`
Code where it raises the issue:
`var attachedMessage = await notificationItem.EncryptedContent.DecryptAsync<ChatMessage>((id, thumbprint) => Task.FromResult(certificate));`

**Reply by mohammadamer** [reply]
@AsmaaNoh
I'll submit my code soon, Just wanted to finish everything before submission :)

**Reply by mohammadamer** [reply]
![image](https://user-images.githubusercontent.com/19314043/225330635-180d7256-1117-431d-bd55-a03472c6c499.png)

**Reply by mohammadamer** [reply]
Here is the [URL](https://github.com/mohammadamer/GraphDotNetHack/blob/main/MSGraph.MeetingRecordingNotifier/MeetingRecordingNotifier/Functions/Notify.cs) for the code of azure function that has the issue.
